src/Gaffer/Switch.cpp

```diff
@@ -34,25 +34,360 @@
 //
 //////////////////////////////////////////////////////////////////////////
 
-#include "Gaffer/Switch.inl"
+#include "Gaffer/Switch.h"
 
-namespace Gaffer
-{
+#include "Gaffer/ArrayPlug.h"
+#include "Gaffer/Context.h"
+#include "Gaffer/ContextAlgo.h"
+#include "Gaffer/MetadataAlgo.h"
+
+#include "boost/bind.hpp"
+#include "boost/bind/placeholders.hpp"
+
+using namespace Gaffer;
 
-// Specialisation of the constructor for when the base
-// class is ComputeNode. This is needed because ComputeNode
-// doesn't have the ArrayPlug constructor that we use in the
-// standard constructor in Switch.inl.
-template<>
-Switch<ComputeNode>::Switch( const std::string &name )
+IE_CORE_DEFINERUNTIMETYPED( Switch );
+
+size_t Switch::g_firstPlugIndex = 0;
+
+Switch::Switch( const std::string &name)
 	:	ComputeNode( name )
 {
-	init( /* expectBaseClassPlugs = */ false );
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	addChild( new IntPlug( "index", Gaffer::Plug::In, 0, 0 ) );
+	addChild( new BoolPlug( "enabled", Gaffer::Plug::In, true ) );
+
+	childAddedSignal().connect( boost::bind( &Switch::childAdded, this, ::_2 ) );
+	plugSetSignal().connect( boost::bind( &Switch::plugSet, this, ::_1 ) );
+	plugInputChangedSignal().connect( boost::bind( &Switch::plugInputChanged, this, ::_1 ) );
+}
+
+Switch::~Switch()
+{
+}
+
+void Switch::setup( const Plug *plug )
+{
+	if( inPlugs() )
+	{
+		throw IECore::Exception( "Switch already has an \"in\" plug." );
+	}
+	if( outPlug() )
+	{
+		throw IECore::Exception( "Switch already has an \"out\" plug." );
+	}
+
+	PlugPtr inElement = plug->createCounterpart( "in0", Plug::In );
+	MetadataAlgo::copyColors( plug , inElement.get() , /* overwrite = */ false  );
+	inElement->setFlags( Plug::Dynamic | Plug::Serialisable, true );
+	ArrayPlugPtr in = new ArrayPlug(
+		"in",
+		Plug::In,
+		inElement,
+		0,
+		Imath::limits<size_t>::max(),
+		Plug::Default | Plug::Dynamic
+	);
+	addChild( in );
+
+	PlugPtr out = plug->createCounterpart( "out", Plug::Out );
+	out->setFlags( Plug::Dynamic | Plug::Serialisable, true );
+	MetadataAlgo::copyColors( plug , out.get() , /* overwrite = */ false  );
```

Comment:
One more thing that lead to the confusion with that odd production example was that our custom pane wasn't displaying the expected results, even when the plug did have a valid value. This is because there is some metadata on the plug used to setup the Switch which tells that pane to interpret the string as a database url. Since the colors were copied over, they appear visually to still be "asset management plugs" but they've lost the other metadata necessary to make them actually behave that way... is there a particular reason we don't copy all the metadata during setup? Maybe there is some metadata that would be bad to copy over?

Reply:
I don't remember to be honest - I know we omit "layout:*" when promoting plugs because it'd be really annoying otherwise, but I don't recall why we went with inclusions here rather than exclusions. My preference would be to consider that as a separate topic, since this is a big enough change already. If that's OK, are you expecting anything else from me before we can merge?

Reply:
That's fine. And no, I think we're ready.